packages/fuels-contract/src/execution_script.rs

```diff
@@ -114,10 +119,136 @@ impl ExecutableFuelCall {
             .iter()
             .any(|r|
                 matches!(r, Receipt::ScriptResult { result, .. } if *result != ScriptExecutionResult::Success)
-        ) {
+            ) {
             return Err(Error::RevertTransactionError(Default::default(), receipts));
         }
 
         Ok(receipts)
     }
+
+    fn extract_duplicate_ids<'a, T: Iterator<Item = &'a Input>>(inputs: T) -> Vec<String> {
+        inputs
+            .filter_map(|input| {
+                input
+                    .utxo_id()
+                    .map(|id| id.tx_id())
+                    .map(Either::Left)
+                    .or_else(|| input.message_id().map(Either::Right))
+            })
+            .duplicates()
+            .map(|id| match id {
+                Either::Left(id) => format!("{id:?}"),
+                Either::Right(id) => format!("{id:?}"),
+            })
+            .collect()
+    }
+
+    fn validate_no_duplicate_ids<'a, T: IntoIterator<Item = &'a Input>>(
+        inputs: T,
+    ) -> Result<(), Error> {
+        let duplicate_ids = Self::extract_duplicate_ids(inputs.into_iter());
+
+        if duplicate_ids.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::InvalidData(format!(
+                "Provided inputs are not unique! Duplicated ids: {duplicate_ids:?}"
+            )))
+        }
+    }
+
+    pub fn set_inputs(
+        &mut self,
+        inputs: Vec<Input>,
+    ) -> std::result::Result<&mut ExecutableFuelCall, Error> {
+        Self::validate_no_duplicate_ids(&inputs)?;
+        *self.tx.inputs_mut() = inputs;
+        Ok(self)
+    }
+
+    pub fn get_inputs(&mut self) -> &Vec<Input> {
+        self.tx.inputs()
+    }
+
+    pub fn extend_inputs(&mut self, inputs: Vec<Input>) -> Result<&mut ExecutableFuelCall, Error> {
+        Self::validate_no_duplicate_ids(chain!(self.tx.inputs(), &inputs))?;
+        self.tx.inputs_mut().extend(inputs);
+        Ok(self)
+    }
```

Comment:
We might have a problem here.

When we request `get_asset_inputs_for_amount` from the node, we'll get resources of the node's choosing following a randomized algorithm to prevent dust accumulation. 

During the formation of our tx we ask for resources to cover the gas fee and cover whatever assets the contract calls required. The node is free to return whichever resources it deems appropriate to cover our request.

Later on, after we've formed the tx, somebody might want to add some more inputs for whatever reason. They're probably going to do that using `extend_inputs`. And they're probably going to get those inputs by calling `get_asset_inputs_for_amount`. 

There is nothing (AFAIK) preventing the possibility that the node is going to return resources that we've already 'reserved' by either covering the transaction costs or because a contract call demanded them. The node can't know we're planning to do anything with the resources it had suggested just a few moments before while we were building the tx.

So a user might find this snippet to work most of the time:

```rust
let mut a_call = contract.methods().some_method().get_executable_call();
let some_more_inputs = wallet.get_asset_inputs_for_amount(...);
a_call.extend_inputs(some_more_inputs);
```
but once in a while, the node will return a resource we've already placed in our tx inputs and then this code will error out due to the duplicated ids.

How often this happens would depend on the client's algorithm for choosing resources.

It would be unfortunate for the user to have his application error-out even though he had the resources to cover his request, but the node choose an already 'reserved' resource to avoid dust.

We could either:
1.  leave the user to his own devices and have him request all his resources and filter out the ones we've already chosen for his tx, or repeat `get_asset_inputs_for_amount` until the node changes its mind, or, 
2. develop a way of extending inputs such that we request resources from the node only once for a given asset id.
3. both of the above, let the user use the error-prone one if he knows what he's doing

The #1 way is more flexible because the user would choose *exactly* which resource he wants to extend as inputs. I'm out of my depth here but that could be used to make decisions in the contract? Heard about predicates making decisions based on inputs.

The #2 way is more friendly, just have the user extend the inputs by noting his wish of `(asset_id, amount)` and we'll find the coins for that wish and for our other needs at the same time, avoiding duplicates. 

So something like:
```rust
call.add_input_resources(asset_id, amount);
```

and right before executing the script, we would do one pass per asset id to actually find all the needed resources.


Dunno. Been doing my own thing for a while now, might be out of the loop. @FuelLabs/sdk-rust opinions?

Reply:
You're right. I will take this into account and try to implement something